src/components/encrypted-intent.ts

```diff
@@ -0,0 +1,237 @@
+import { UserMembership } from "./membership-cache";
+import { APPSERVICE_LOGIN_TYPE, ClientEncryptionSession } from "./encryption";
+import Logging from "./logging";
+import { ReadStream } from "fs";
+import BotSdk, { MatrixClient } from "matrix-bot-sdk";
+import { FileUploadOpts, Intent, IntentOpts } from "./intent";
+
+const log = Logging.get("EncryptedIntent");
+
+export interface EncryptedIntentOpts {
+    sessionPromise: Promise<ClientEncryptionSession|null>;
+    sessionCreatedCallback: (session: ClientEncryptionSession) => Promise<void>;
+    ensureClientSyncingCallback: () => Promise<void>;
+    originalHomeserverUrl: string;
+}
+
+/**
+ * Implements some special handling on top of Intent to handle encrypted rooms.
+ */
+export class EncryptedIntent extends Intent {
+
+    private readonly encryptedRooms = new Map<string, string|false>();
```

Comment:
Do we care what algo it is? It is in the public API technically, but named as if it was a bool. Not sure how much of a breaking change that'd be, but I'd break it to rename it, if it's needed at all :)

Reply:
Fair, we can save some bytes here.